plumbing/bootstrap.py

```diff
@@ -74,88 +67,79 @@ def clean_error_message(message):
 
     with open(description_file) as stream:
         try:
             package = yaml.load(stream.read())
             package['name'] = slugify(package['title'], separator='_')
             package['extraction_processors'] = select_processors(package)
-        except ParserError as error:
+        except (ParserError, ScannerError) as error:
             message = clean_error_message(str(error))
             package = {'yaml_error': message}
             logging.debug('Parser error in %s: %s', description_file, message)
 
     with open(datapackage_file, 'w+') as stream:
         stream.write(json.dumps(package, indent=4))
 
-    logging.debug('Created %s', datapackage_file)
+    pipeline_id = os.path.split(source_folder)[-1]
+    logging.info('%s: created datapackage', pipeline_id)
 
     return package
 
 
-def bootstrap_pipeline(source_folder):
+def bootstrap_pipeline(package, source_folder):
     """Save the pipeline file inside the source folder."""
 
     _, source_name = os.path.split(source_folder)
-    package = load_source_datapackage(source_folder)
     pipeline = {source_name: DEFAULT_PIPELINE}
 
     if 'yaml_error' not in package:
         processors = select_processors(package)
-        save_processor_placeholders(processors)
+        save_processor_placeholders(processors, source_folder)
         save_pipeline(pipeline, source_folder)
 
 
 def save_pipeline(pipeline, source_folder):
-    filepath = os.path.join(source_folder, PIPELINE_SPECS)
+    filepath = os.path.join(source_folder, PIPELINE_FILE)
     with open(filepath, 'w+') as stream:
         stream.write(yaml.dump(pipeline))
-    logging.debug('Created %s', filepath)
+    pipeline_id = os.path.split(source_folder)[-1]
+    logging.info('%s: created pipeline specs', pipeline_id)
 
 
 def select_processors(package):
     """Select the appropriate processors for the pipeline."""
 
-    mode = package['extraction_mode']
+    for resource in package['resources']:
+        mode = resource['extraction_mode']
 
-    if sum(map(mode.values(), int)) > 1:
-        message = '{}: extraction mode is ambiguous'
-        raise ValueError(message.format(package['name']))
+        if sum(map(int, mode.values())) > 1:
+            message = '{}: extraction mode is ambiguous'
+            raise ValueError(message.format(package['name']))
 
-    if mode['download_link']:
-        return ['download_remote_sources']
-    elif mode['web_scraper']:
-        return ['scrape_remote_sources']
-    elif mode['pdf_scraper']:
-        return ['download_pdf_sources', 'scrape_pdf_sources']
-    elif mode['user_interface']:
-        raise NotImplementedError('User interface mode not ready yet.')
-    else:
-        ValueError('No extraction mode specified in the description file.')
+        if any([mode['web_scraper'],
+               mode['pdf_scraper'],
+               mode['user_interface']]):
+            return ['scraper']
+        else:
+            return []
```

Comment:
@pwalsh Okay I just caught a proper design mistake here in the my source.description.yaml specs. The processors must be defined at the datapackage level not the resource level. So for now, as a work-around, i'll just bootstrap the processors from the 1st resource. The description specs will need to be changed.